### PR TITLE
[CORE UPDATE - PART IV] Fix cryptography+cffi

### PR DESCRIPTION
**Pull request by opacam**
This is one more part for the pr #1537 (discussed previously in #1460 and #1534)

Here I fix recipes cryptography and cffi (and touch some others which those depend on) to make it work with the core-update pull request, which also contains a new version of the openssl libs created by @tito. I tried to update the most of the packages to latest version and grants compatibility for both versions of python.

Note: this has to be merged only when the core-update has been merged because it depends on the new python2 recipe and the openssl libs (1.1)

### CODE REVIEW

**Comment by JonasT** [comment]
This is another one that is quite fundamentally relevant for everyone, isn't it? Doesn't ctypes require cffi? Once this is in I can do the `ctypes.util.find_library()` patch on top, and finally all the core things will work properly out of the box! :heart_eyes:

**Comment by JonasT** [comment]
Ok this **DOESN'T** work for me. The reason is that `cffi` has `setuptools` in the dependencies, and `setuptools` expects `ctypes` to be already around, which apparently for some reason it isn't (maybe because libffi isn't there yet? Not sure):

```
[INFO]:    Building setuptools for armeabi-v7a
[INFO]:    setuptools apparently isn't already in site-packages
[INFO]:    Installing setuptools into site-packages
[INFO]:    -> directory context /root/.local/share/python-for-android/build/other_builds/setuptools-python3/armeabi-v7a__ndk_target_21/setuptools
[DEBUG]:   -> running python setup.py install -O2 --root=/root/.local/share/python-for-android/build/python-installs/unnamed_dist_1 --install-lib=.
[DEBUG]:   	Traceback (most recent call last):
[DEBUG]:   	  File "setup.py", line 11, in <module>
[DEBUG]:   	    import setuptools
[DEBUG]:   	  File "/root/.local/share/python-for-android/build/other_builds/setuptools-python3/armeabi-v7a__ndk_target_21/setuptools/setuptools/__init__.py", line 14, in <module>
[DEBUG]:   	    from setuptools.dist import Distribution, Feature
[DEBUG]:   	  File "/root/.local/share/python-for-android/build/other_builds/setuptools-python3/armeabi-v7a__ndk_target_21/setuptools/setuptools/dist.py", line 25, in <module>
[DEBUG]:   	    from setuptools import windows_support
[DEBUG]:   	  File "/root/.local/share/python-for-android/build/other_builds/setuptools-python3/armeabi-v7a__ndk_target_21/setuptools/setuptools/windows_support.py", line 2, in <module>
[DEBUG]:   	    import ctypes
[DEBUG]:   	  File "/root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Lib/ctypes/__init__.py", line 7, in <module>
[DEBUG]:   	    from _ctypes import Union, Structure, Array
[DEBUG]:   	ModuleNotFoundError: No module named '_ctypes
```
(this is on latest p4a git master with this pull request on op, with `cffi` and `python3` in the `--requirements` list)

**Comment by JonasT** [comment]
I tried adding `libffi` as `depends` for both `python3` and `setuptools`, and bumping setuptools to `40.6.3`, but nothing helped, sadly. Still stuck with the same error

**Comment by misl6** [comment]
Worked for me. (on top of the current master).

Just added libffi as first requirement in buildozer.
`requirements = libffi, openssl, python3, kivy==master, android, cython, pyjnius, pycryptodome`

**Comment by JonasT** [comment]
Ok, I spotted this output: `[DEBUG]:   	checking for --with-system-ffi... yes` - but still no `_ctypes` module. I will try enabling openssl now and see if that is connected (which I think it shouldn't be)

**Comment by JonasT** [comment]
`--requirements` I used: `libffi,openssl,python3,cffi,sdl2,pyjnius,android,Pillow,/my-app/.android-build/code-lib-copy/`

In the hostpython configure output I'm spotting: `[DEBUG]: checking for --with-system-ffi... yes`

However, I still get above error and poked around in the build folder afterwards:

```
# /root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/python
Python 3.7.1 (default, Jan 10 2019, 21:28:13) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Lib/ctypes/__init__.py", line 7, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
>>> import ssl
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Lib/ssl.py", line 98, in <module>
    import _ssl             # if we can't import it, let the error propagate
ModuleNotFoundError: No module named '_ssl'
>>>
```
I don't know the exact folder layout very well, but I'd also expect `_ctypes.so` in here:
```
# ls /root/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/Modules/
Setup            _collectionsmodule.o  _math.o      _threadmodule.o  bufferedio.o  errnomodule.o   getbuildinfo.o  itertoolsmodule.o  pwdmodule.o       textio.o
Setup.local      _functoolsmodule.o    _operator.o  _tracemalloc.o   bytesio.o     faulthandler.o  getpath.o       ld_so_aix          signalmodule.o    timemodule.o
_abc.o           _iomodule.o           _sre.o       _weakref.o       config.c      fileio.o        hashtable.o     main.o             stringio.o        xxsubtype.o
_codecsmodule.o  _localemodule.o       _stat.o      atexitmodule.o   config.o      gcmodule.o      iobase.o        posixmodule.o      symtablemodule.o  zipimport.o
```
... and there's nothing like that in the entire folder structure:
```
# find /root/.local/share/python-for-android/build/other_builds/hostpython3/ | grep 'ctypes.so'
#
```

**Comment by JonasT** [comment]
Ah lol. I don't have `libffi-dev` on the system itself, and given this is **host**python that's probably it, right? I'll install it and try again. :woman_facepalming:

**Comment by opacam** [comment]
Yes, you need libffi-dev for hostpython3

**Comment by opacam** [comment]
Aaa...I see, it's above...

**Comment by JonasT** [comment]
I just realized I installed `libffi-dev` in the wrong environment :joy: hang on. I am testing too many things today I'm confusing myself :woman_facepalming: lol

**Comment by opacam** [comment]
No problem, let me know if you need help, today I'm going to bed :sleeping:
...when those things happen it's a signal  :joy: 
 :first_quarter_moon_with_face:

**Comment by JonasT** [comment]
:+1: things also built for me now (with `python 3`) so that's good!

File: pythonforandroid/recipes/cffi/__init__.py [hunk 1]
```diff
@@ -44,15 +43,10 @@ def get_recipe_env(self, arch=None):
...
-            # only keeps major.minor (discards patch)
-            python_version = self.ctx.python_recipe.version[0:3]
-            ndk_dir_python = os.path.join(self.ctx.ndk_dir, 'sources/python/', python_version)
-            env['LDFLAGS'] += ' -L{}'.format(os.path.join(ndk_dir_python, 'libs', arch.arch))
-            env['LDFLAGS'] += ' -lpython{}m'.format(python_version)
-            # until `pythonforandroid/archs.py` gets merged upstream:
-            # https://github.com/kivy/python-for-android/pull/1250/files#diff-569e13021e33ced8b54385f55b49cbe6
-            env['CFLAGS'] += ' -I{}/include/python/'.format(ndk_dir_python)
+        env['LDFLAGS'] += ' -L{}'.format(self.ctx.python_recipe.link_root(arch.arch))
+        env['LDFLAGS'] += ' -lpython{}'.format(self.ctx.python_recipe.major_minor_version_string)
```
[the remark this answers is not in the document]
Comment (AndreMiras) [reply pythonforandroid/recipes/cffi/__init__.py]: That's a pretty valid concern @JonasT and definitely deserves some investigations. In the meantime this PR is already improving the current state while not bringing any regression to me.
I think we could merge and eventually address that later

[the remark this answers is not in the document]
Comment (JonasT) [reply pythonforandroid/recipes/cffi/__init__.py]: :+1:

**Pull request merged**